### PR TITLE
add RISC-V github CI & update chipStar to v1.2 + LLVM 19

### PR DESCRIPTION
**Pull request by franz**

### CODE REVIEW

File: lib/llvmopencl/EmitPrintf.cc [hunk 1]
```diff
@@ -427,18 +432,33 @@ static Value *processNonStringArg(Value *Arg, IRBuilder<> &Builder) {
...
+
+  // float cast is disabled; not all devices support fp64
+/*
   if (Ty->isFloatingPointTy()) {
     if (DL.getTypeAllocSize(Ty) < 8) {
       return Builder.CreateFPExt(Arg, Builder.getDoubleTy());
     }
   }
-
+*/
```
Comment (linehill) [inline lib/llvmopencl/EmitPrintf.cc:461]: Dead code?

File: lib/llvmopencl/linker.cpp [hunk 2]
```diff
@@ -451,6 +488,58 @@ static void shared_copy(llvm::Module *program, const llvm::Module *lib,
...
+// can be called with pointer arguments with different AS based
+// on compilation input (OpenCL C source vs SPIR-V).
+//
+// Here, we replace the calls directly with address
+// space casts. This works for uniform address space targets (CPUs),
+// while the disjoint AS targets can override the implementation
+// as needed.
+static bool convertAddrSpaceOperator(llvm::Function *Func, std::string &Log) {
+  if (Func == nullptr) {
+    //Log.append("convertASoperator : Function not found\n");
```
Comment (linehill) [inline lib/llvmopencl/linker.cpp:503]: Dead code?

File: lib/llvmopencl/linker.cpp [hunk 3]
```diff
@@ -451,6 +488,58 @@ static void shared_copy(llvm::Module *program, const llvm::Module *lib,
...
+    return true;
+  }
+
+  if (!Func->isDeclaration()) {
+    // If the builtin library provides an implementation, don't touch it
+    //Log.append("convertASoperator : Function is not declaration\n");
```
Comment (linehill) [inline lib/llvmopencl/linker.cpp:509]: Dead code?

File: lib/llvmopencl/linker.cpp [hunk 4]
```diff
@@ -647,83 +732,61 @@ int link(llvm::Module *Program, const llvm::Module *Lib, std::string &Log,
...
-              Call->eraseFromParent();
-            }
-            }
-          }
-          continue;
-        }
 
-        if ((f == NULL) || (f->isDeclaration() &&
+  if (Program->getTargetTriple().compare(0,5,"nvptx") != 0) {
+  for (auto &DeclIter : DeclaredFunctions) {
```
Comment (linehill) [inline lib/llvmopencl/linker.cpp:760]: Indentation.

File: lib/llvmopencl/EmitPrintf.cc [hunk 5]
```diff
@@ -326,7 +326,12 @@ callBufferedPrintfStart(IRBuilder<> &Builder, SmallVector<llvm::Value *> &Args,
             StringData(StringRef(), LenWithNull, LenWithNullAligned, false));
       }
     } else {
-      TypeSize AllocSize = DL.getTypeAllocSize(Args[i]->getType());
+      Type *T = Args[i]->getType();
+      TypeSize AllocSize = DL.getTypeAllocSize(T);
+      // stores of Ints&Vectors <= 64bits are expanded to 64bits
+      if (T->isIntOrIntVectorTy() && (DL.getTypeSizeInBits(T) < 64)) {
+        AllocSize = TypeSize::get(8, false);
+      }
```
Comment (linehill) [inline lib/llvmopencl/EmitPrintf.cc:334]: LLVM style nit:
```suggestion
      if (T->isIntOrIntVectorTy() && (DL.getTypeSizeInBits(T) < 64))
        AllocSize = TypeSize::get(8, false);
```

File: doc/sphinx/source/notes_6_1.rst [hunk 6]
```diff
@@ -39,6 +39,12 @@ Work-group vectorization improvements
 Miscellaneous
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
+* New env variable POCL_MAX_COMPUTE_UNITS. This is similar to
+  existing POCL_CPU_MAX_CU_COUNT, but is not driver-specific,
+  hence can be used to limit the CU count of all drivers that
+  support limiting CU count. If both are specified, the driver
+  specific variable takes precedence.
+
```
Comment (linehill) [inline doc/sphinx/source/notes_6_1.rst:47]: Is there use cases for `POCL_CPU_MAX_CU_COUNT`? Could all uses of it replaced with the`POCL_CPU_MAX_COMPUTE_UNITS`? I was thinking if the `POCL_CPU_MAX_CU_COUNT` should be deprecated (document it, emit warning message if it used).

**Review by linehill — APPROVED** [review]
LGTM.

File: doc/sphinx/source/notes_6_1.rst (hunk 6 above)
Comment (pjaaskel) [reply doc/sphinx/source/notes_6_1.rst:47]: I think the use case is exactly that: If we only want to limit the CPU's max compute unit count, but not other drivers. I don't know if this is a realistic use case, and we discussed this locally and came to the conclusion that we have so many envs that it doesn't hurt to keep this one ;)

**Pull request merged**